File: Assignments/A04/3.cpp
#include <iostream>
#include <vector>

using namespace std;

int longest_increasing_subsequence(vector<int>& a, int n) {
    vector<int> lis{a[0]};

    for (int i = 1; i < n; i++) {
        if (lis.back() >= a[i]) {
            int l = 0, r = lis.size()-1;
            while (l <= r) {
                int m = (l + r) / 2;
                if (lis[m] < a[i]) {
                    l = m + 1;
                } else {
                    r = m - 1;
                }
            }
            lis[l] = a[i];
        } else {
            lis.push_back(a[i]);
        }
    }

    return lis.size();
}

int main() {
    int n;
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    cout << longest_increasing_subsequence(a, n) << endl;

    return 0;
}

File: Assignments/A04/3.py
def longest_increasing_subsequence(a, n):
    lis = [a[0]]

    for i in range(1, n):
        if lis[-1] >= a[i]:
            l, r = 0, len(lis)-1
            while l <= r:
                m = int((l + r) / 2)
                if lis[m] < a[i]:
                    l = m + 1
                else:
                    r = m - 1
            lis[l] = a[i]
        else:
            lis.append(a[i])

    return len(lis)

n = int(input())
a = list(map(int, input().split()))
print(longest_increasing_subsequence(a, n))
